File: frontend/src/lib/posts.ts
import { apiFetch } from './api'
import { getAccessToken } from './auth'

export type Post = {
  id: string
  title: string
  slug: string
  excerpt: string | null
  cover_image_url: string | null
  published_at: string | null
  updated_at: string | null
  created_at?: string
  content_json?: Record<string, unknown>
  content_html?: string | null
}

export async function listPublishedPosts() {
  return apiFetch<Post[]>('/posts', { method: 'GET' })
}

export async function getPostById(id: string) {
  const token = getAccessToken()
  return apiFetch<Post>(`/posts/${id}`, { method: 'GET', token })
}

export type PostUpsertInput = {
  title: string
  slug?: string | null
  excerpt?: string | null
  cover_image_url?: string | null
  content_json: Record<string, unknown>
  content_html: string
  published: boolean
}

export async function createPost(input: PostUpsertInput): Promise<Post> {
  const token = getAccessToken()
  if (!token) throw new Error('No autenticado')
  return apiFetch<Post>('/posts', { method: 'POST', token, body: input })
}

export async function updatePost(postId: string, input: PostUpsertInput): Promise<Post> {
  const token = getAccessToken()
  if (!token) throw new Error('No autenticado')
  return apiFetch<Post>(`/posts/${postId}`, { method: 'PATCH', token, body: input })
}

export async function deletePost(postId: string): Promise<void> {
  const token = getAccessToken()
  if (!token) throw new Error('No autenticado')
  await apiFetch(`/posts/${postId}`, { method: 'DELETE', token })
}

export async function uploadImageToSupabase(file: File) {
  const token = getAccessToken()
  if (!token) throw new Error('No autenticado')

  const url = new URL((import.meta.env.VITE_API_URL as string) || 'http://localhost:8000')
  url.pathname = '/uploads/image'

  const form = new FormData()
  form.append('file', file)

  const res = await fetch(url.toString(), {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${token}`,
    },
    body: form,
  })

  if (!res.ok) {
    const msg = await res.text().catch(() => '')
    throw new Error(`Upload falló: ${res.status} ${msg}`)
  }

  return (await res.json()) as { url: string; path?: string }
}

File: backend/app/routes/posts.py
import re
from datetime import datetime
from typing import Any

from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel

from ..supabase_http import rest_delete, rest_insert, rest_patch, rest_select


router = APIRouter(prefix="/posts", tags=["posts"])


def _get_bearer_token(request: Request) -> str | None:
    auth = request.headers.get("authorization")
    if not auth:
        return None
    parts = auth.split()
    if len(parts) != 2 or parts[0].lower() != "bearer":
        return None
    return parts[1]


def _slugify(value: str) -> str:
    value = value.strip().lower()
    value = re.sub(r"[^a-z0-9]+", "-", value)
    value = re.sub(r"-{2,}", "-", value)
    return value.strip("-")


class PostUpsertBody(BaseModel):
    title: str
    slug: str | None = None
    excerpt: str | None = None
    cover_image_url: str | None = None
    content_json: dict[str, Any]
    content_html: str
    published: bool = False


@router.get("")
async def list_published(request: Request):
    try:
        # Solo “publicadas”: mantenemos el filtro en backend (aunque RLS puede también restringir).
        payload = await rest_select(
            "posts",
            access_token=None,
            params={
                "select": "id,title,slug,excerpt,cover_image_url,published_at,updated_at",
                "order": "updated_at.desc",
                "published_at": "not.is.null",
            },
        )
        return payload
    except Exception as e:
        raise HTTPException(status_code=502, detail=f"Error consultando Supabase (/posts): {e}")


@router.get("/{post_id}")
async def get_post(post_id: str, request: Request):
    token = _get_bearer_token(request)

    try:
        payload = await rest_select(
            "posts",
            access_token=token,
            params={
                "select": "id,title,slug,excerpt,cover_image_url,published_at,created_at,updated_at,content_json,content_html",
                "id": f"eq.{post_id}",
                "limit": 1,
            },
        )
    except Exception as e:
        raise HTTPException(status_code=502, detail=f"Error consultando Supabase (/posts/{post_id}): {e}")

    if not payload:
        raise HTTPException(status_code=404, detail="Post no encontrado")
    return payload[0]


@router.post("")
async def create_post(body: PostUpsertBody, request: Request):
    token = _get_bearer_token(request)
    if not token:
        raise HTTPException(status_code=401, detail="No autenticado")

    slug = body.slug or _slugify(body.title)
    published_at = datetime.utcnow().isoformat() + "Z" if body.published else None

    payload = await rest_insert(
        "posts",
        access_token=token,
        payload=[
            {
                "title": body.title,
                "slug": slug,
                "excerpt": body.excerpt,
                "cover_image_url": body.cover_image_url,
                "content_json": body.content_json,
                "content_html": body.content_html,
                "published_at": published_at,
            }
        ],
    )
    return payload[0] if isinstance(payload, list) and payload else payload


@router.patch("/{post_id}")
async def update_post(post_id: str, body: PostUpsertBody, request: Request):
    token = _get_bearer_token(request)
    if not token:
        raise HTTPException(status_code=401, detail="No autenticado")

    slug = body.slug or _slugify(body.title)
    published_at = datetime.utcnow().isoformat() + "Z" if body.published else None

    payload = await rest_patch(
        "posts",
        access_token=token,
        params={"id": f"eq.{post_id}"},
        payload={
            "title": body.title,
            "slug": slug,
            "excerpt": body.excerpt,
            "cover_image_url": body.cover_image_url,
            "content_json": body.content_json,
            "content_html": body.content_html,
            "published_at": published_at,
        },
    )
    return payload[0] if isinstance(payload, list) and payload else payload


@router.delete("/{post_id}")
async def delete_post(post_id: str, request: Request):
    token = _get_bearer_token(request)
    if not token:
        raise HTTPException(status_code=401, detail="No autenticado")

    payload = await rest_delete("posts", access_token=token, params={"id": f"eq.{post_id}"})
    return {"deleted": True, "result": payload}

